### PR TITLE
Mlgym#247 gpt 2

### LINKED ISSUE #247: Implementation of GPT-2 in MLGym

### PR DESCRIPTION
**Pull request by moinam**
Creating pull request for GPT 2 integration with ML Gym.

Note:
1) ~Perplexity loss is still under progress so I have not committed the final changes required to do so.~
2) ~I have made changes to the preprocessing part which has not been committed yet since they are in progress.~
3) ~The changes in the pull request are without runtime bugs so you can train it if you want and can reproduce the same scenario as what I have shown in the meetings.~

~GPT 2 integration is complete for mlgym, hugging face - pytorch with DDP  and accelerate.~

~TODO: Revert to correct logging of events in mlgym.~

Ready to merge.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -134,6 +134,20 @@ cd event_storage/2022-11-06--17-59-10/
...
+
+### Creating Blueprint
+
+### Using custom Neural net model
+
+### Creating custom activation, loss, metric functions
+
+### Using a local dataset
+
 ## MLboard
```
Comment (le1nux) [inline README.md:151]: After merging @vijulshah's PR we should reference the MLBoard tutorial here.

File: example/gpt2/huggingface_impl/accelerate_run_exp.sh [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+CUDA_VISIBLE_DEVICES='1,2,3,4' accelerate launch --multi_gpu --gpu_ids 1,2,3,4 --num_processes=4 /cluster/home/mlgym/example/gpt2/huggingface_impl/gpt2_hug_impl_accelerate.py
+
+CUDA_VISIBLE_DEVICES='1,2,3,4,5' accelerate launch --multi_gpu --gpu_ids 1,2,3,4,5 --num_processes=5 /cluster/home/mlgym/example/gpt2/huggingface_impl/gpt2_hug_impl_accelerate.py
+
+CUDA_VISIBLE_DEVICES='1,2,3,4,5,6' accelerate launch --multi_gpu --gpu_ids 1,2,3,4,5,6 --num_processes=6 /cluster/home/mlgym/example/gpt2/huggingface_impl/gpt2_hug_impl_accelerate.py
+
+CUDA_VISIBLE_DEVICES='1,2,3,4,5,6,7' accelerate launch --multi_gpu --gpu_ids 1,2,3,4,5,6,7 --num_processes=7 /cluster/home/mlgym/example/gpt2/huggingface_impl/gpt2_hug_impl_accelerate.py
+
+CUDA_VISIBLE_DEVICES='0,1,2,3,4,5,6,7' accelerate launch --multi_gpu --gpu_ids 0,1,2,3,4,5,6,7 --num_processes=8 /cluster/home/mlgym/example/gpt2/huggingface_impl/gpt2_hug_impl_accelerate.py
```
Comment (le1nux) [inline example/gpt2/huggingface_impl/accelerate_run_exp.sh:15]: this we don't need to merge to master

File: example/gpt2/huggingface_impl/gpt2_hug_impl.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
+import datetime
```
Comment (le1nux) [inline example/gpt2/huggingface_impl/gpt2_hug_impl.py:1]: we don't use this file within MLgym right?

File: example/gpt2/huggingface_impl/gpt2_hug_impl_accelerate.py [hunk 4]
```diff
@@ -0,0 +1,272 @@
+import datetime
```
Comment (le1nux) [inline example/gpt2/huggingface_impl/gpt2_hug_impl_accelerate.py:1]: we don't use this file within MLgym right?

File: example/gpt2/huggingface_impl/gpt2_hug_impl_ddp.py [hunk 5]
```diff
@@ -0,0 +1,365 @@
+import datetime
```
Comment (le1nux) [inline example/gpt2/huggingface_impl/gpt2_hug_impl_ddp.py:1]: we don't use this file within MLgym right?

File: example/gpt2/huggingface_impl/run_experiments.sh [hunk 6]
```diff
@@ -0,0 +1,17 @@
+#!/bin/sh
```
Comment (le1nux) [inline example/gpt2/huggingface_impl/run_experiments.sh:1]: we don't use this file within MLgym right?

File: example/gpt2/huggingface_impl/viz_gpt2_hug_impl.ipynb
Comment (le1nux) [inline example/gpt2/huggingface_impl/viz_gpt2_hug_impl.ipynb:1]: I think we also don't need this in master. AFAIK, we only used this to make the evaluations for the paper

File: icde_paper/notebooks/runtime_evaluation.ipynb
Comment (le1nux) [inline icde_paper/notebooks/runtime_evaluation.ipynb:1]: We only needed this for the paper. So I wouldn't merge it to master.

**Review by le1nux — CHANGES_REQUESTED** [review]
Looks good to me! :-) Also, thanks for taking the time to add comments to the parts that were lacking documentation. 
I left a few comments. Afterwards we can merge. 👍

File: README.md (hunk 1 above)
Comment (moinam) [reply README.md:151]: Ok I understand.

File: example/gpt2/huggingface_impl/accelerate_run_exp.sh (hunk 2 above)
Comment (moinam) [reply example/gpt2/huggingface_impl/accelerate_run_exp.sh:15]: yes

File: example/gpt2/huggingface_impl/gpt2_hug_impl.py (hunk 3 above)
Comment (moinam) [reply example/gpt2/huggingface_impl/gpt2_hug_impl.py:1]: Will remove the whole folder of huggingface_impl in the gpt2 example. That was only done for paper. No need in master :)

**Review by le1nux — APPROVED** [review]
LGTM :-) Great work! Merging 👍
